tools/functional-tester/etcd-agent/agent.go

```diff
@@ -75,23 +77,60 @@ func (a *Agent) start(args ...string) error {
 	return nil
 }
 
+var alreadyFinished = errors.New("os: process already finished")
+
 // stop stops the existing etcd process the agent started.
 func (a *Agent) stop() error {
 	if a.state != stateStarted {
 		return nil
 	}
-	err := a.cmd.Process.Kill()
+
+	err := waitCmdExit(a.cmd, syscall.SIGTERM)
 	if err != nil {
 		return err
 	}
-	_, err = a.cmd.Process.Wait()
+
+	err = a.cmd.Process.Kill()
+	if err.Error() == alreadyFinished.Error() {
+		a.state = stateStopped
+		err = nil
+	}
+	return err
+}
+
+var sigToState = map[syscall.Signal]string{
+	syscall.SIGTERM: "signal: terminated",
+}
+
+func waitCmdExit(cmd *exec.Cmd, sig syscall.Signal) error {
+	err := cmd.Process.Signal(sig)
 	if err != nil {
 		return err
-
 	}
 
-	a.state = stateStopped
-	return nil
+	done, errc := make(chan struct{}), make(chan error)
+	go func() {
+		st, err := cmd.Process.Wait()
+		if err != nil {
+			errc <- err
+		}
+		if st.Exited() {
+			done <- struct{}{}
+		}
+		if st.String() == sigToState[sig] {
+			done <- struct{}{}
+		}
+	}()
+
+	select {
+	case <-done:
+		return nil
+	case err := <-errc:
+		return err
+	case <-time.After(5 * time.Second):
+		cmd.Process.Kill()
+		return fmt.Errorf("timed out")
```

Comment:
this still needs to know the `Wait()` completed-- an `<-errc` will work if the `done` stuff goes away